### ISSUE TITLE
Reference external data with backtesting

### ISSUE THREAD

**Issue by darioneto** [freqtrade/freqtrade#7347]
Operating system: Docker
Python Version: 3.10.6_ 
CCXT version: _1.92.52 
Freqtrade Version: __ docker-a6d78a86__
  
I'm not sure what is the problem or if possible to reference from the external data source for the backtesting purpose. 
The same works fine with the populating entries/exit but when I try backtesting there are no hits. 
Backtesting returns 0 trades. Live trades are fine on the same queries.
The time schema and format are the same as in the freqtrade backtesting 1 min.

**Comment by darioneto** [comment]
looks like I have more luck with embedding the output inside of the populate_indicators functions. Not sure if this function is required for backtesting to be holding the data. However there, I came across a '['open', 'high', 'low', 'close'] not in index"' error. Do you know if there are any specific settings to be changed to integrate with external data? I think this is related to the time column compatibility.
```
2022-09-05 11:17:19,641 - freqtrade.configuration.load_config - INFO - Using config: user_data/config.json ...
2022-09-05 11:17:19,642 - freqtrade.loggers - INFO - Verbosity set to 0
2022-09-05 11:17:19,642 - freqtrade.configuration.configuration - INFO - Using max_open_trades: 1 ...
2022-09-05 11:17:19,651 - freqtrade.configuration.configuration - INFO - Using user-data directory: /freqtrade/user_data ...
2022-09-05 11:17:19,651 - freqtrade.configuration.configuration - INFO - Using data directory: /freqtrade/user_data/data/binance ...
2022-09-05 11:17:19,651 - freqtrade.configuration.configuration - INFO - Parameter --cache=day detected ...
2022-09-05 11:17:19,651 - freqtrade.configuration.check_exchange - INFO - Checking exchange...
2022-09-05 11:17:19,668 - freqtrade.configuration.check_exchange - INFO - Exchange "binance" is officially supported by the Freqtrade development team.
2022-09-05 11:17:19,668 - freqtrade.configuration.configuration - INFO - Using pairlist from configuration.
2022-09-05 11:17:19,669 - freqtrade.configuration.config_validation - INFO - Validating configuration ...
2022-09-05 11:17:19,673 - freqtrade.commands.optimize_commands - INFO - Starting freqtrade in Backtesting mode
2022-09-05 11:17:19,673 - freqtrade.exchange.exchange - INFO - Instance is running with dry_run enabled
2022-09-05 11:17:19,674 - freqtrade.exchange.exchange - INFO - Using CCXT 1.92.20
2022-09-05 11:17:19,674 - freqtrade.exchange.exchange - INFO - Applying additional ccxt config: {'options': {'defaultType': 'future'}}
2022-09-05 11:17:19,689 - freqtrade.exchange.exchange - INFO - Applying additional ccxt config: {'options': {'defaultType': 'future'}}
2022-09-05 11:17:19,705 - freqtrade.exchange.exchange - INFO - Using Exchange "Binance"
2022-09-05 11:17:20,621 - freqtrade.resolvers.exchange_resolver - INFO - Using resolved exchange 'Binance'...
2022-09-05 11:17:20,706 - freqtrade.resolvers.iresolver - INFO - Using resolved strategy bot_eth_long_short_strategy from '/freqtrade/user_data/strategies/eth_strategy_short_test.py'...
2022-09-05 11:17:20,706 - freqtrade.strategy.hyper - INFO - Found no parameter file.
2022-09-05 11:17:20,707 - freqtrade.resolvers.strategy_resolver - INFO - Override strategy 'stake_currency' with value in config file: USDT.
2022-09-05 11:17:20,707 - freqtrade.resolvers.strategy_resolver - INFO - Override strategy 'stake_amount' with value in config file: 200.
2022-09-05 11:17:20,707 - freqtrade.resolvers.strategy_resolver - INFO - Override strategy 'unfilledtimeout' with value in config file: {'entry': 10, 'exit': 10, 'exit_timeout_count': 0, 'unit': 'minutes'}.
2022-09-05 11:17:20,707 - freqtrade.resolvers.strategy_resolver - INFO - Strategy using minimal_roi: {'0': 0.025}
2022-09-05 11:17:20,707 - freqtrade.resolvers.strategy_resolver - INFO - Strategy using timeframe: 1m
2022-09-05 11:17:20,707 - freqtrade.resolvers.strategy_resolver - INFO - Strategy using stoploss: -0.035
2022-09-05 11:17:20,707 - freqtrade.resolvers.strategy_resolver - INFO - Strategy using trailing_stop: False
2022-09-05 11:17:20,707 - freqtrade.resolvers.strategy_resolver - INFO - Strategy using trailing_stop_positive_offset: 0.0
2022-09-05 11:17:20,707 - freqtrade.resolvers.strategy_resolver - INFO - Strategy using trailing_only_offset_is_reached: False
2022-09-05 11:17:20,707 - freqtrade.resolvers.strategy_resolver - INFO - Strategy using use_custom_stoploss: True
2022-09-05 11:17:20,707 - freqtrade.resolvers.strategy_resolver - INFO - Strategy using process_only_new_candles: False
2022-09-05 11:17:20,707 - freqtrade.resolvers.strategy_resolver - INFO - Strategy using order_types: {'entry': 'limit', 'exit': 'limit', 'stoploss': 'market', 'stoploss_on_exchange': True}
2022-09-05 11:17:20,707 - freqtrade.resolvers.strategy_resolver - INFO - Strategy using order_time_in_force: {'entry': 'gtc', 'exit': 'gtc'}
2022-09-05 11:17:20,708 - freqtrade.resolvers.strategy_resolver - INFO - Strategy using stake_currency: USDT
2022-09-05 11:17:20,708 - freqtrade.resolvers.strategy_resolver - INFO - Strategy using stake_amount: 200
2022-09-05 11:17:20,708 - freqtrade.resolvers.strategy_resolver - INFO - Strategy using protections: []
2022-09-05 11:17:20,708 - freqtrade.resolvers.strategy_resolver - INFO - Strategy using startup_candle_count: 30
2022-09-05 11:17:20,708 - freqtrade.resolvers.strategy_resolver - INFO - Strategy using unfilledtimeout: {'entry': 10, 'exit': 10, 'exit_timeout_count': 0, 'unit': 'minutes'}
2022-09-05 11:17:20,708 - freqtrade.resolvers.strategy_resolver - INFO - Strategy using use_exit_signal: True
2022-09-05 11:17:20,708 - freqtrade.resolvers.strategy_resolver - INFO - Strategy using exit_profit_only: False
2022-09-05 11:17:20,708 - freqtrade.resolvers.strategy_resolver - INFO - Strategy using ignore_roi_if_entry_signal: False
2022-09-05 11:17:20,708 - freqtrade.resolvers.strategy_resolver - INFO - Strategy using exit_profit_offset: 0.0
2022-09-05 11:17:20,708 - freqtrade.resolvers.strategy_resolver - INFO - Strategy using disable_dataframe_checks: False
2022-09-05 11:17:20,708 - freqtrade.resolvers.strategy_resolver - INFO - Strategy using ignore_buying_expired_candle_after: 0
2022-09-05 11:17:20,708 - freqtrade.resolvers.strategy_resolver - INFO - Strategy using position_adjustment_enable: True
2022-09-05 11:17:20,708 - freqtrade.resolvers.strategy_resolver - INFO - Strategy using max_entry_position_adjustment: -1
2022-09-05 11:17:20,709 - freqtrade.configuration.config_validation - INFO - Validating configuration ...
2022-09-05 11:17:20,722 - freqtrade.resolvers.iresolver - INFO - Using resolved pairlist StaticPairList from '/freqtrade/freqtrade/plugins/pairlist/StaticPairList.py'...
2022-09-05 11:17:20,724 - freqtrade.data.history.history_utils - INFO - Using indicator startup period: 30 ...
2022-09-05 11:17:20,822 - freqtrade.optimize.backtesting - INFO - Loading data from 2022-08-06 08:17:00 up to 2022-09-05 08:14:00 (29 days).
2022-09-05 11:17:20,822 - freqtrade.configuration.timerange - WARNING - Moving start-date by 30 candles to account for startup time.
2022-09-05 11:17:20,829 - freqtrade.data.history.idatahandler - WARNING - ETH/USDT, funding_rate, 8h, data starts at 2022-08-06 16:00:00
2022-09-05 11:17:20,845 - freqtrade.data.history.idatahandler - WARNING - ETH/USDT, mark, 8h, data starts at 2022-08-06 16:00:00
2022-09-05 11:17:20,857 - freqtrade.optimize.backtesting - INFO - Dataload complete. Calculating indicators
2022-09-05 11:17:20,858 - freqtrade.optimize.backtesting - WARNING - Backtest result caching disabled due to use of open-ended timerange.
2022-09-05 11:17:20,858 - freqtrade.optimize.backtesting - INFO - Running backtesting for Strategy bot_eth_long_short_strategy
2022-09-05 11:17:20,858 - freqtrade.strategy.hyper - INFO - No params for buy found, using default values.
2022-09-05 11:17:20,858 - freqtrade.strategy.hyper - INFO - No params for sell found, using default values.
2022-09-05 11:17:20,859 - freqtrade.strategy.hyper - INFO - No params for protection found, using default values.
                        date  pcnt_change_down    round
0  2022-09-05 11:15:00+00:00                 0  0.14687
1  2022-09-05 11:12:00+00:00                 0  0.14687
2  2022-09-05 11:09:00+00:00                 0  0.14687
3  2022-09-05 11:05:00+00:00                 0  0.14687
4  2022-09-05 11:03:00+00:00                 0  0.14693
..                       ...               ...      ...
95 2022-09-05 06:29:00+00:00                 0  0.14600
96 2022-09-05 06:26:00+00:00                 0  0.14600
97 2022-09-05 06:23:00+00:00                 0  0.14600
98 2022-09-05 06:21:00+00:00                 0  0.14595
99 2022-09-05 06:17:00+00:00                 0  0.14595

[100 rows x 3 columns]
2022-09-05 11:17:20,953 - freqtrade.optimize.backtesting - INFO - Backtesting with data from 2022-09-05 06:17:00 up to 2022-09-05 09:44:00 (0 days).
                        date  pcnt_change_down    round  enter_short enter_tag
95 2022-09-05 06:29:00+00:00                 0  0.14600          NaN       NaN
96 2022-09-05 06:26:00+00:00                 0  0.14600          NaN       NaN
97 2022-09-05 06:23:00+00:00                 0  0.14600          NaN       NaN
98 2022-09-05 06:21:00+00:00                 0  0.14595          NaN       NaN
99 2022-09-05 06:17:00+00:00                 0  0.14595          NaN       NaN
                        date  pcnt_change_down    round  enter_short enter_tag  exit_short exit_tag
95 2022-09-05 06:29:00+00:00                 0  0.14600          NaN       NaN         NaN      NaN
96 2022-09-05 06:26:00+00:00                 0  0.14600          NaN       NaN         NaN      NaN
97 2022-09-05 06:23:00+00:00                 0  0.14600          NaN       NaN         NaN      NaN
98 2022-09-05 06:21:00+00:00                 0  0.14595          NaN       NaN         NaN      NaN
99 2022-09-05 06:17:00+00:00                 0  0.14595          NaN       NaN         NaN      NaN
2022-09-05 11:17:20,976 - freqtrade - ERROR - Fatal exception!
Traceback (most recent call last):
  File "/freqtrade/freqtrade/main.py", line 37, in main
    return_code = args['func'](args)
  File "/freqtrade/freqtrade/commands/optimize_commands.py", line 58, in start_backtesting
    backtesting.start()
  File "/freqtrade/freqtrade/optimize/backtesting.py", line 1315, in start
    min_date, max_date = self.backtest_one_strategy(strat, data, timerange)
  File "/freqtrade/freqtrade/optimize/backtesting.py", line 1229, in backtest_one_strategy
    results = self.backtest(
  File "/freqtrade/freqtrade/optimize/backtesting.py", line 1081, in backtest
    data: Dict = self._get_ohlcv_as_lists(processed)
  File "/freqtrade/freqtrade/optimize/backtesting.py", line 388, in _get_ohlcv_as_lists
    data[pair] = df_analyzed[HEADERS].values.tolist() if not df_analyzed.empty else []
  File "/home/ftuser/.local/lib/python3.10/site-packages/pandas/core/frame.py", line 3511, in __getitem__
    indexer = self.columns._get_indexer_strict(key, "columns")[1]
  File "/home/ftuser/.local/lib/python3.10/site-packages/pandas/core/indexes/base.py", line 5782, in _get_indexer_strict
    self._raise_if_missing(keyarr, indexer, axis_name)
  File "/home/ftuser/.local/lib/python3.10/site-packages/pandas/core/indexes/base.py", line 5845, in _raise_if_missing
    raise KeyError(f"{not_found} not in index")
KeyError: "['open', 'high', 'low', 'close'] not in index"
```

**Comment by xmatthias** [comment]
using non-freqtrade data as main source (for pricing data) is not supported.
The data does not seem to contain regular candle data (ohlcv) - or they're named wrong.
This may work in live as we're not using these columns there (however freqUI will not be able to display graphs with this).

Either way - you shouldn't be doing it this way.

If the date format matches - then you should merge the additional data to the dataframe freqtrade provides (treat your data as some sort of "informative data" - the docs do contain samples for that), and use that going forward.

**Comment by darioneto** [comment]
the additional data is only meant for the trading signals which I was planning to use for backtesting. I figured it sort of works  inside of the populte_indicators, but I then I was facing the problem with "date" first, thus I swapped the time column to date as a workaround, however it seems like there are some other issues such as 
"Dataframe returned from strategy has mismatching length" which effectively causes KeyError: "['open', 'high', 'low', 'close'] not in index"
What I would like to do is to attach additional columns inside of a dataframe that can match/join ON the existing timestamp/date from freqtrade and then use those fields for the backtesting. I don't need to substitute internal pricing data, just add some more to the existing dataframe.  Graphs are the least important as long as I can get metrics from backtesting in raw format via cli.
Would this be possible? if not any alternative method you could suggest? informative data as a workaround?

**Comment by xmatthias** [comment]
while there's no explicit samples for this in the docs - you should still be able to use [merge_informative_pair](https://www.freqtrade.io/en/stable/strategy-customization/#merge_informative_pair) to merge your dataframe to the original dataframe.

alternatively you can also use pd.merge directly (use pandas docs for details) - the correct merge method here will depend a bit on the meaning of your time values (is it candle start date - or "information available date").

**Comment by darioneto** [comment]
the matching would be based on time. Also, there will be more than one table to merge (10 - 20) depending on the strategy. Thus I'm not certain first how merge_informative_pair() will be suitable as I don't see any detailed info in the docs. Looks like pd.merge would do, just not sure about the performance hit. Maybe better to merge those all into one big table before injecting them to freqtrade for better performance.

**Comment by xmatthias** [comment]
Merging pandas dataframes is usually quick.


You'll have to be very careful with dates. the date column in freqtrade's dataframe is the candle OPEN time - so the "completed" time is actually shifted by one candle.
`merge_informative_pair` takes care of this (it's also just pd.merge in the end - with some quirks applied to account for the "open date" problem) - but without knowing your source, it's difficult to determine what your date columns correspond to.
As mentioned, documentation will not cover your case specifically, but treating your data as "informative" dataframe should make most of the docs apply to your case, too.

**Comment by darioneto** [comment]
the strategy is not candlestick but time, thus the first match will be good. From what I understand I should use the pd.merge in my case. My data is pretty simple time type columns in 1 min format (same as freqtrade) and the rest are all numeric columns.

So when I go with this statement joining 2 data frames

**`dataframe = pd.merge(dataframe, dataframe_2, left_on='date', how='left')`**

 I get an Error:
 Must pass "right_on" OR "right_index"
 then I add right_on='date'

`**dataframe = pd.merge(dataframe, dataframe_2, left_on='date',right_on='date', how='left')**`

I still get the same error "Dataframe returned from strategy has mismatching length"

However, from the logs looks like it's merging it together on the Left Join

```
bot_eth_short  | 2022-09-05 13:26:12,699 - freqtrade.strategy.interface - WARNING - Unable to analyze candle (OHLCV) data for pair ETH/USDT: Dataframe returned from strategy has mismatching length.
bot_eth_short  | 2022-09-05 13:26:17,708 - freqtrade.strategy.interface - WARNING - Unable to analyze candle (OHLCV) data for pair ETH/USDT: Dataframe returned from strategy has mismatching length.
bot_eth_short  | 2022-09-05 13:26:22,737 - freqtrade.strategy.interface - WARNING - Unable to analyze candle (OHLCV) data for pair ETH/USDT: Dataframe returned from strategy has mismatching length.
bot_eth_short  | 2022-09-05 13:26:27,717 - freqtrade.strategy.interface - WARNING - Unable to analyze candle (OHLCV) data for pair ETH/USDT: Dataframe returned from strategy has mismatching length.
bot_eth_short  |                          date     open  ...  pcnt_change_down    round
bot_eth_short  | 0   2022-09-04 20:46:00+00:00  1571.35  ...               NaN      NaN
bot_eth_short  | 1   2022-09-04 20:47:00+00:00  1569.90  ...               NaN      NaN
bot_eth_short  | 2   2022-09-04 20:48:00+00:00  1569.10  ...               NaN      NaN
bot_eth_short  | 3   2022-09-04 20:49:00+00:00  1568.54  ...               NaN      NaN
bot_eth_short  | 4   2022-09-04 20:50:00+00:00  1568.92  ...               NaN      NaN
bot_eth_short  | ..                        ...      ...  ...               ...      ...
bot_eth_short  | 994 2022-09-05 13:20:00+00:00  1575.06  ...               NaN      NaN
bot_eth_short  | 995 2022-09-05 13:21:00+00:00  1575.73  ...               0.0  0.14900
bot_eth_short  | 996 2022-09-05 13:22:00+00:00  1575.87  ...               NaN      NaN
bot_eth_short  | 997 2022-09-05 13:23:00+00:00  1578.15  ...               NaN      NaN
bot_eth_short  | 998 2022-09-05 13:24:00+00:00  1577.48  ...               0.0  0.14927
bot_eth_short  | 
bot_eth_short  | [999 rows x 8 columns]
bot_eth_short  |                          date     open  ...  pcnt_change_down    round
bot_eth_short  | 0   2022-09-04 20:46:00+00:00  1571.35  ...               NaN      NaN
bot_eth_short  | 1   2022-09-04 20:47:00+00:00  1569.90  ...               NaN      NaN
bot_eth_short  | 2   2022-09-04 20:48:00+00:00  1569.10  ...               NaN      NaN
bot_eth_short  | 3   2022-09-04 20:49:00+00:00  1568.54  ...               NaN      NaN
bot_eth_short  | 4   2022-09-04 20:50:00+00:00  1568.92  ...               NaN      NaN
bot_eth_short  | ..                        ...      ...  ...               ...      ...
bot_eth_short  | 994 2022-09-05 13:20:00+00:00  1575.06  ...               NaN      NaN
bot_eth_short  | 995 2022-09-05 13:21:00+00:00  1575.73  ...               0.0  0.14900
bot_eth_short  | 996 2022-09-05 13:22:00+00:00  1575.87  ...               NaN      NaN
bot_eth_short  | 997 2022-09-05 13:23:00+00:00  1578.15  ...               NaN      NaN
bot_eth_short  | 998 2022-09-05 13:24:00+00:00  1577.48  ...               0.0  0.14927
bot_eth_short  | 
bot_eth_short  | [999 rows x 8 columns]

```

The above output is what I would expect as the time will only match in periodic 1m intervals, not a constant stream.

**Comment by xmatthias** [comment]
> Dataframe returned from strategy has mismatching length

this simply means that you're doing "something" to the dataframe that's changing it's length.
you get a 1000 length dataframe in - and you're now missing one  line (for example) - or there's one additional line for whatever reason.

**Comment by darioneto** [comment]
just to clarify the pd.merge should be done inside of populate_indicators , informative_pairs function or it doesn't matter?

**Comment by xmatthias** [comment]
`populate_indicators()`.

as you have only a simple date - i suspect that you'll be introducing a lookahead bias with the way you're merging (even though it's only looking ahead for 1 minute).
in 90% of the cases, the date associated with a datapoint is the date when this information is available (so the earliest point you can look at data is at 10:00).
If you merge this to a 1h candle at 10:00, you're actually looking at a candle that's only available at 11:00 - so the date should be shifted by one interval.

**Comment by darioneto** [comment]
ok looks like it was working since the first merge, Bcktesting works also. I just realized that the entry/exits were still using wrong references and it was coming from the entry trigger, not the merged table. 
So my intention is to generate a signal as soon as the match with new data in a table is found. Not waiting on candle close etc. It should be triggered as soon as it validates conditions from external DB and then some few seconds delay untill freqtrade wil pick it up working of the queue. Do you suggest the initial idea of the merge_informative_pair for this?

**Comment by xmatthias** [comment]
it's your data - so you must know / be aware if your data requires a shift to avoid future-looking data or not.

To properly think this through - switch to 1h candles (it's easier on the mind to think in hours than in minutes - in minutes it's dangerous to think "It's just a few seconds ...") - and think about the process of a candle that opens (that's the date we get) - shows high/low - and then closes (that's the date we get the candle in freqtrade - but the label is on the open).
Now think when your signal is generated. If it's based on ANYTHING other than the candle's open price (like high or low) - then you'll have to do the shift.

**Comment by darioneto** [comment]
Great, thanks for your sugestions. Do you mind to share, what the freqtrade compatible "shift" syntax looks like if I want to work with pd.merge which already works. 
`dataframe = pd.merge(dataframe, dataframe_2, left_on='date',right_on='date', how='left')`

**Comment by xmatthias** [comment]
Simply look at the function that implements this...

https://github.com/freqtrade/freqtrade/blob/3f3099cbfcbb2188c232d23dee940c615744f347/freqtrade/strategy/strategy_helper.py#L6-L72

**Comment by darioneto** [comment]
looks complex :-), seems it would be easier to configure merge_informative_pair as you mentioned earlier. 
Do you mind sharing the compatible with freqtrade syntax of informative_pair example equivalent to the below statement
`dataframe = pd.merge(dataframe, dataframe_2, left_on='date',right_on='date', how='left')`

**Comment by xmatthias** [comment]
well - like the one i [linked to above](https://www.freqtrade.io/en/stable/strategy-customization/#merge_informative_pair) ? 
there's a full sample available a bit [further down](https://www.freqtrade.io/en/stable/strategy-customization/#complete-data-provider-sample) that doc page, too.

**Comment by darioneto** [comment]
ok ok, I missed that bottom part, It makes sense now and works fine. Much appreciate your help on this

**Comment by darioneto** [comment]
my last question on this subject is how could I return multiple dataframes inside of informative_pairs and then merge them together inside populate_indicators function?

**Comment by xmatthias** [comment]
you don't - but you can call informative_pairs multiple times.

**Comment by darioneto** [comment]
So the limitation here is: I can only return one dataframe per function in which case is only one inside of "informative_pairs" function. However with pd.merge I could potentially do as many as the resources allow? Is that correct?

**Comment by xmatthias** [comment]
No.
Please look at the [full sample](https://www.freqtrade.io/en/stable/strategy-customization/#complete-data-provider-sample) page. 

`dataframe = merge_informative_pair(dataframe, informative, self.timeframe, inf_tf, ffill=True)` is just one call - you can call this 
however many times - for each of the additional "informative" dataframes you've got (just make sure that dataframe is always passed on).

`informative_pairs` is really irrelevant - as you'll not be using informative pairs - but some custom data from wherever - so you'll just use the "Merging" code.

**Comment by darioneto** [comment]
ok from what I understood earlier I need to place each of the additional "informative" dataframes  inside of informative_pairs function. That was a reason why I was asking if multiple dataframes can be returned. 
So according to the last message I can define all of it  inside of a populate_indicators function then I understand this is can be called multiple times.

**Comment by xmatthias** [comment]
Please read that whole documentation page. It's all explained there - including a full sample.
It'll require some imagination on how to apply that to your usecase - but without that - i doubt you'll succeed in implementing this.

**Comment by darioneto** [comment]
my imagination is much larger than my knowledge of programming ;-), I will figure it out at some point, will get it tested probably the fastest way to find out, thanks anyways

**Comment by darioneto** [comment]
ok after some testing it looks like it works just fine. 
Just on a side note, is it possible to reference dataframe or use enter_tags within the custom stop_loss function?

**Comment by xmatthias** [comment]
Documentation.

**Comment by darioneto** [comment]
I'm not sure if this bug or an expected error with multiple merges. When I merge multiple dataframes at some point I get an error 
`cannot reindex on an axis with duplicate labels`

 print(dataframe.index.duplicated()) returns 
```
bot_btc  |  False False False False False False False False False False False False
bot_btc  |  False False False False False False False False False False False False
bot_btc  |  False False False False False False False False False False False False
bot_btc  |  False False False False False False False False False False False False
bot_btc  |  False False False False False False False False False False False False
bot_btc  |  False False False False False False False False False False False False
bot_btc  |  False False False False False False False False False False False False
bot_btc  |  False False False False False False False False False False False False
bot_btc  |  False False False False False False False False False False False False
```

I've found the that only way to fix it is to change the timeframe on the matching dataframe 

for example, if I use different times for each merge statement it will work fine ie: 

```
      inf_up = '1m'
      inf_down = '1m'
      inf_bo = '2m'
      inf_pc60min = '1m'
      inf_pdown_sens = '3m'
```
The weird thing is that the first two are the same and that also works, but only for the first two. The moment I add another one with 1 min it starts to complain. 

tried to reset index with dataframe = dataframe.reset_index()
but then I get another error :cannot insert level_0, already exists"

my intention is to match based on the same timeframe of 1m. Would you know what is the problem related to? 
What is also interesting is that there is no error with pd.merge method, only merge_informative_pair is having an issue while merging.

**Comment by xmatthias** [comment]
the error points to multiple columns being named the same. You're not really checking for that above.

**Comment by darioneto** [comment]
I read the docs many times over at this point and can't figure out why this is happening. I know for a fact that the columns' names are different. Only date column is the same across all dataframes, but it has to be, to be able to join. 
If the column's names would be overlapping then the pd.merge method would also fail I'd imagine, but it works with the same dataframes.

According to the documentation 
        _# Use the helper function merge_informative_pair to safely merge the pair
        # Automatically renames the columns and merges a shorter timeframe dataframe and a longer timeframe informative pair_
In theory, this should rename the columns automatically. 
I tried to force to rename before the merge with `informative.columns = [f"{col}_{inf_tf}" for col in informative.columns]` but then it causes another issue with date column and unable to join.
```
  |   File "pandas/_libs/index.pyx", line 136, in pandas._libs.index.IndexEngine.get_loc
bot_btc  |   File "pandas/_libs/index.pyx", line 163, in pandas._libs.index.IndexEngine.get_loc
bot_btc  |   File "pandas/_libs/hashtable_class_helper.pxi", line 5198, in pandas._libs.hashtable.PyObjectHashTable.get_item
bot_btc  |   File "pandas/_libs/hashtable_class_helper.pxi", line 5206, in pandas._libs.hashtable.PyObjectHashTable.get_item
bot_btc  | KeyError: 'date'
```

**Comment by xmatthias** [comment]
your easiest way to debug this is by printing the columns after each merge step.

You clearly can't rename all columns like this (not date, anyway) - but if you run `merge` multiple times - you need to keep in mind how renaming works - so if you have 2 identical columns - on the identical timeframe - they'll end up the same.
Pandas won't really care about this up to a certain degree (for example renaming an index) - so the error is usually not where the problem happens - but when it's actually causing you problems.

**Comment by darioneto** [comment]
Yeah, I printed those earlier, and one by one they all pass. however when I enable a set of those frames it fails. One weird merge in isolation prints `date_1m` for some reason.  I think I've got the same problem as in this case discussed here: ((https://github.com/freqtrade/freqtrade/issues/7001))

Anyways I've changed everything already to pd.merge as that worked with no issues and looks more straightforward to understand.

I remember you mentioned earlier about the lookahead and now I think I understand how this concept works while debugging. 
Easier to understand on example rather than in abstract statement. 
So the latest timestamp in fact is delayed one period in reference to the chosen strategy timeframe. Thus, I must shift the results from dataframes to match the time in freqtrade based on that concept. For example, if my strategy is set to "1m"
 then the dataframe date match (for the fastest execution) should be set to the latest minute - 1. Is that correct?

**Comment by xmatthias** [comment]
dataframe dates are candle open dates - the close time of a candle is the next date in the dataframe.

**Comment by darioneto** [comment]
then seems correct then, when the dataframe from external reference was updated with a limit of one index, which is a present minute. That is the future from the freqtrade perspective. Therefore some of the signals were never triggered (with exception of some of them where count > 1)
As for the live trading mode do you recommend to set as .iloc[-1] for dataframes while populating entries/exits to be sure the timestamp is matched as the earliest update?
